solr/solr-ref-guide/src/dense-vector-search.adoc

```diff
@@ -0,0 +1,308 @@
+= Dense Vector Search
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+The Apache Solr *Dense Vector Search* module adds support for indexing and searching dense numerical vectors.
+
+https://en.wikipedia.org/wiki/Deep_learning[Deep learning] can be used to produce a vector representation of both the query and the documents in a corpus of information.
+
+These neural network-based techniques are usually referred to as neural search, an industry derivation from the academic field of https://www.microsoft.com/en-us/research/uploads/prod/2017/06/fntir2018-neuralir-mitra.pdf[Neural information Retrieval].
+
+== Important Concepts
+
+=== Dense Vector Representation 
+A dense vector describes information as an array of elements, each of them explicitly defined.
```

Comment:
my intention was to clarify the distinction between dense and sparse vectors, I have just refined the doc, let me know if it's any better.
Feel free to recommend any addition and I'll add it!
As a non-native speaker/writer, I am finding this feedback on the documentation extremely valuable, I am really grateful!

Reply:
Ahh I see! Sorry for the delayed response. I'll take a crack at this, feel free to take or leave this suggestion:

```
=== Dense Vector Representation 
A traditional "tokenized inverted index" can be considered to model text as a "sparse" vector, in which each term in the index corresponds to one vector dimension. In such a model, the number of dimensions is generally quite high (corresponding to term cardinality), and the vector for any given document contains mostly "zeros" (hence it is "sparse", as only a handful of terms that exist in the overall index will be present in any given document).

"Dense vector" representation contrasts with term-based "sparse vector" representation in that it distills semantic meaning into a fixed (and limited) number of dimensions. The number of dimensions in this approach is generally much lower than the "sparse" case, and the vector for any given document is "dense", as most of its dimensions are populated by non-zero values.

Solr exposes the capability to index and search dense vectors; but in contrast to the "sparse" approach (for which Solr provides tokenizers to "generate sparse vectors" directly from text input) the task of _generating_ vectors must be handled in application logic external to Solr. There may be cases where it makes sense to directly search data that natively exists as a vector (e.g., scientific data); but in a "text search" context, it is likely that users will leverage deep learning models such as https://en.wikipedia.org/wiki/BERT_(language_model)[BERT] to encode textual information as dense vectors, supplying the resulting vectors to Solr explicitly at index and query time.
```

Reply:
Thanks again @magibney for the feedback, I gave it a go!
I incorporated your suggestions (with some minor changes here and there), hopefully, it looks all right now!

Reply:
Thanks, Alessandro!

I think maybe we should go either/or with reference to "Bag of words" vs. "Inverted index". In the text I drafted I referred to an "inverted index" as an example of sparse vector representation because I figured it would likely be the most familiar reference point for the "Solr refguide" audience.

I note that Wikipedia has two separate entries: [Bag-of-words model](https://en.wikipedia.org/wiki/Bag-of-words_model) and [Inverted index](https://en.wikipedia.org/wiki/Inverted_index), neither of which refers directly to the other, but both of which include a "See also" reference to the page for [Vector space model](https://en.wikipedia.org/wiki/Vector_space_model). The wikipedia pages are explicit about one difference, with the second clearly describing how most people use Solr:
1. "The Bag-of-words model is an orderless document representation — only the counts of words matter."
2. "A word-level inverted index (or full inverted index or inverted list) additionally contains the positions of each word within a document.[2] The latter form offers more functionality (like phrase searches)"

Either model is arguably a valid example of a "sparse vector" representation. My practical concern is that in this context, it would be easy to misinterpret the reference to "Bag of words" as an implicit reference to "_non-vector_ search in Solr", whatever the intention of using "Bag-of-words" as a point of reference. I think that interpretation would be misleading, and could obscure the true distinction between "dense vector search" and classic TF-IDF/BM25/phrase-boosted/etc. search in Solr, and the appropriate use cases for each approach.

(minor point: I'd also be inclined to drop the separate heading for "Sparse Vector Representation" -- I was purposefully vague in saying `can be considered to model text as a "sparse" vector`; perhaps I'm mistaken, but my impression is that "sparse" retrieval models (as _accurate_ as that characterization may be), are most often characterized as "sparse" as a foil to explain why "dense retrieval" is characterized as "dense". Indeed, that's what we're doing here!)

Reply:
I agree with your observations!
From a practical perspective, how do you want me to change the current documentation piece in your opinion?

Reply:
Thanks! I took another pass at this below.

Starting from your latest revision, I partially reverted a few things to the original version I drafted (mostly just the beginning of the first paragraph). I stuck with your reformatting into more separate paragraphs, and I appreciated your reduction of my "overuse" of "quotes" :-) -- so left that as you had it.

I also left the two places where you replaced "Solr" with "Apache Solr", though personally I'd be inclined to go with the more succinct "Solr"? I linked to the wikipedia page for "inverted index", analogous to what you had done for "Bag of words" -- though I'm not sure the external link is necessary.

The only other change I introduced was "distills semantic meaning" => "distills _approximate_ semantic meaning". I don't feel too strongly about that, but I do think it's more accurate, and lets us humans continue to feel special :-)
 
>=== Dense Vector Representation 
>A traditional tokenized https://en.wikipedia.org/wiki/Inverted_index[inverted index] can be considered to model text as a "sparse" vector, in which each term in the corpus corresponds to one vector dimension. In such a model, the number of dimensions is generally quite high (corresponding to the term dictionary cardinality), and the vector for any given document contains mostly zeros (hence it is sparse, as only a handful of terms that exist in the overall index will be present in any given document).
>
>Dense vector representation contrasts with term-based sparse vector representation in that it distills approximate semantic meaning into a fixed (and limited) number of dimensions.
>
>The number of dimensions in this approach is generally much lower than the sparse case, and the vector for any given document is dense, as most of its dimensions are populated by non-zero values.
>
>In contrast to the sparse approach (for which tokenizers are used to generate sparse vectors directly from text input) the task of generating vectors must be handled in application logic external to Apache Solr.
>
>There may be cases where it makes sense to directly search data that natively exists as a vector (e.g., scientific data); but in a text search context, it is likely that users will leverage deep learning models such as https://en.wikipedia.org/wiki/BERT_(language_model)[BERT] to encode textual information as dense vectors, supplying the resulting vectors to Apache Solr explicitly at index and query time.
>
>For additional information you can refer to this https://sease.io/2021/12/using-bert-to-improve-search-relevance.html[blog post].

Reply:
That's brilliant! I just committed the changes!
Resolving now!